README.md
# JWST User Toolkit

Python libraries and scripts use James Webb Space Telescope (JWST) users.

Documentation:
- [Installation](https://github.com/spacetelescope/jwstuser/wiki/Installation) gives installation instructions

Contributing:
- Use [Sphinx-RTD](https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html) for docstrings.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

# [0.0.2] - 2022-05-03
### Added
- More definitions for SQL-to-python data type conversions.
- Attribute to `timeseries` class that returns times in `MJD` as well (`self.times_mjd`). 
- Created this `CHANGELOG.md`.
- Created the `tutorials/get_fgs_centroids.py` script that showcases usage for a real-world application.
### Changed
- Bumped version number.

# [0.0.1] - 2022-03-18
### Added
- Repository; initial version of the code.


src/jwstuser/util.py
from getpass import getpass
from pathlib import Path
from os import getenv

def get_mast_api_token(mast_api_token=None):
    '''Get MAST API token. Precedence is arg, env, file, prompt.'''
    if not mast_api_token:
        mast_api_token = getenv('MAST_API_TOKEN')
    if not mast_api_token:
        path = Path.home() / '.mast_api_token'
        try:
            with open(path, 'r') as fp:
                lines = fp.read().splitlines()
                if len(lines) == 1:
                    mast_api_token = lines[0]
                else:
                    print('Ignoring ~/.mast_api_token, expected one line')
        except FileNotFoundError:
            pass
    if not mast_api_token and prompt:
        mast_api_token = getpass('Enter MAST API token: ')
    try:
        return verified_mast_api_token(mast_api_token)
    except ValueError as e:
        raise e

def verified_mast_api_token(token):
    '''Return input MAST API token, if type, length, characters are valid.'''
    is_not = 'MAST API token is not'
    if token:
        if type(token) is str:
            if len(token) == 32:
                if token.isalnum():
                    return token
                else:
                    raise ValueError(f"{is_not} alphanumeric: '{token}'")
            else:
                raise ValueError(f"{is_not} 32 characters: '{token}'")
        else:
            raise ValueError(f"{is_not} type str: {type(token)}")
    else:
        raise ValueError(f"{is_not} defined")


src/jwstuser/engdb.py
import numpy as np
from collections import namedtuple
from csv import reader as csv_reader
from datetime import datetime
from requests import get as requests_get
from statistics import mode

from .mastapi import get_mast_api_token

class UnauthorizedError(Exception):
    def __init__(self, message):
        super(UnauthorizedError, self).__init__(message)
        self.message = message


class EngineeringDatabase:
    '''Access JWST engineering database hosted by MAST at STScI.'''
    def __init__(self, mast_api_token=None):
        self.token = get_mast_api_token(mast_api_token)
        self.baseurl = 'https://mast.stsci.edu/jwst/api/v0.1/' \
            'Download/file?uri=mast:jwstedb'

    def format_date(self, date):
        '''Convert datetime object or ISO 8501 string to EDB date format.'''
        if type(date) is str:
            dtobj = datetime.fromisoformat(date)
        elif type(data) is datetime:
            dtobj = date
        else:
            raise ValueError('date must be ISO 8501 string or datetime obj')
        return dtobj.strftime('%Y%m%dT%H%M%S')

    def timeseries(self, mnemonic, start, end):
        '''Get engineering data for specified mnemonic and time interval.'''
        startdate = self.format_date(start)
        enddate = self.format_date(end)
        filename = f'{mnemonic}-{startdate}-{enddate}.csv'
        url = f'{self.baseurl}/{filename}'
        headers = {'Authorization': f'token {self.token}'}
        with requests_get(url, headers=headers, stream=True) as response:
            if response.status_code == 401:
                raise UnauthorizedError('check that MAST API token is valid')
            response.raise_for_status()
            return EdbTimeSeries(mnemonic, response.text.splitlines())

class EdbTimeSeries:
    '''Handle time series data from the JWST engineering database.'''
    def __init__(self, mnemonic, lines):
        self.mnemonic = mnemonic
        self.time, self.time_mjd, self.value = self.parse(lines)
        self._cadence = None
        self._largest_gap = None

    def __len__(self):
        '''Return number of points in time series.'''
        return len(self.time)

    @property
    def timestep_seconds(self):
        '''Return time step between successive times in seconds.'''
        try:
            return [(b - a).total_seconds() for  a, b in zip(
                self.time[:-1], self.time[1:])]
        except IndexError:
            return None

    @property
    def cadence_seconds(self):
        '''Return most common time step in seconds.'''
        timestep_seconds = self.timestep_seconds
        if timestep_seconds:
            self._cadence = mode(timestep_seconds)
            self._largest_gap = max(timestep_seconds)
        return self._cadence

    @property
    def largest_gap_seconds(self):
        '''Return most common time step in seconds.'''
        timestep_seconds = self.timestep_seconds
        if timestep_seconds:
            self._cadence = mode(timestep_seconds)
            self._largest_gap = max(timestep_seconds)
        return self._largest_gap

    def parse(self, lines):
        '''Parse lines of text returned by MAST EDB interface.'''
        # Define python analog of SQL data types.
        # https://docs.microsoft.com/en-us/sql/machine-learning/python
        #     /python-libraries-and-data-types?view=sql-server-ver15
        cast = {'bigint': float, 
                'binary': bytes,
                'bit': bool,
                'char': str,
                'date': datetime,
                'datetime': datetime,
                'float': float, 
                'nchar': str,
                'nvarchar': str,
                'nvarchar(max)': str,
                'real': float,
                'smalldatetime': datetime,
                'smallint': int, 
                'tinyint': int,
                'uniqueidentifier': str,
                'varbinary': bytes,
                'varbinary(max)': bytes,
                'varchar': str, 
                'varchar(n)': str,
                'varchar(max)': str}
        
        # Initialize return variables.
        time = []
        time_mjd = []
        value = []

        for field in csv_reader(lines, delimiter=',', quotechar='"'):

            # Ignore header row.
            if field[0] == 'theTime':
                continue

            # Extract SQL data type in engineering database.
            sqltype = field[3]

            # Convert SQL type to python type. Save time, MJD, and value.
            time.append(datetime.fromisoformat(field[0]))
            time_mjd.append(float(field[1]))
            value.append(cast[sqltype](field[2]))

        return time, time_mjd, value


tutorials/get_fgs_centroids.py
from jwstuser import engdb

# Open a user-defined api_token.txt file containing the MAST API token:
api_token = open('api_token.txt','r').readline()[:-1]

EDB = engdb.EngineeringDatabase(mast_api_token = api_token)

# First, get Guide Star X-position over a pre-determined range of time:
mnemonic = 'SA_ZFGGSPOSX'
start = '2022-05-02T06:00:00'
end = '2022-05-02T13:30:00'

cen_x = EDB.timeseries(mnemonic, start, end)

# Now get Y-position:
mnemonic = 'SA_ZFGGSPOSY'

cen_y = EDB.timeseries(mnemonic, start, end)

# Plot both:
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
sns.set_style('ticks')

plt.plot(cen_x.time, cen_x.value - np.median(cen_x.value), '.-', label = 'Guide Star Position_X', color = 'cornflowerblue')
plt.plot(cen_y.time, cen_y.value - np.median(cen_y.value), '.-', label = 'Guide Star Position_Y', color = 'tomato')

plt.xticks(fontsize = 16)
plt.yticks(fontsize = 16)

plt.title('FGS Centroid Position on May 2nd, 2022')
plt.xlabel('Time (UTC)', fontsize = 20)
plt.ylabel('Position (wrt to median)', fontsize = 20)
plt.legend(fontsize = 18)

plt.show()

# Now extract to an ascii file (first column time in MJD UTC, second column X-position, third column Y-position):
fout = open('fgs_centroids_hp14_may2-2022.dat', 'w')

fout.write('# MJD (UTC; ISIM) \t Centroid X \t Centroid Y\n')
for i in range( len(cen_x.time) ):

    fout.write('{0:.10f} \t {1:.7f} \t {2:.7f}\n'.format( cen_x.time_mjd[i], cen_x.value[i], cen_y.value[i] ))   

fout.close()


src/jwstuser/mastapi.py
from datetime import datetime, timezone
from getpass import getpass
from math import nan
from pathlib import Path
from os import getenv

from astropy.table import unique as table_unique
from astropy.time import Time
from astroquery.mast import Mast, Observations
from numpy import isnan as np_isnan

class JwstFilteredQuery:
    '''Manage MAST API query based on JWST FITS header keyword values.

    Example:
        query = JwstFilteredQuery('Nirspec')
        query.filter_by_values('readpatt', 'NRSRAPID, NRSIRS2RAPID')
        query.filter_by_minmax('nints', 2, 99)
        query.filter_by_timerange('date_beg', '2022-04-02 05:00:00', 59671.8)
        query.append_output_columns('pi_name')
        query.execute_query()
        query.get_caom_product_list()
        query.browse()
    '''
    def __init__(self, collection):
        self._collection = collection
        self._service = f'Mast.Jwst.Filtered.{collection}'
        self.set_output_columns_to_default()
        self.filters = []
        self._params = None
        self.result = None
        self._dataset = None

    @property
    def collection(self):
        '''Return JWST collection specified during instantiation.'''
        return self._collection

    @property
    def service(self):
        '''Return service access point for specified JWST collection.'''
        return self._service

    @property
    def params(self):
        '''Return parameter dictionary for most recent query.'''
        return self._params

    @property
    def dataset(self):
        '''Return dataset names for most recent query.'''
        return self._dataset

    @property
    def caom_obsid(self):
        '''Return CAOM obsid for datasets returned by the most recent query.'''
        return self._caom_obsid

    def filter_by_values(self, keyword, values):
        '''Require keyword value to be in enumerated list.

        Input values may be str with comma-separated values or list of str.
        Whitespace around commas is ignored
        Examples:
            filter_by_values('detector', 'NRS1')
            filter_by_values('detector', 'NRS1,NRS2')
            filter_by_values('detector', 'NRS1, NRS2')
            filter_by_values('detector', ['NRS1'])
            filter_by_values('detector', ['NRS1', 'NRS2'])
        '''
        try:
            valuelist = [v.strip() for v in values.split(',')]
        except AttributeError:
            valuelist = [str(v) for v in values]
        newfilter = {
            'paramName': str(keyword),
            'values': valuelist}
        self.filters.append(newfilter)

    def filter_by_minmax(self, keyword, minval, maxval):
        '''Require keyword value to be in specified range.

        Example:
            filter_by_values('nints', 2, 99)
        '''
        newfilter = {
            'paramName': str(keyword),
            'values': [{'min': minval, 'max': maxval}]}
        self.filters.append(newfilter)

    def filter_by_timerange(self, keyword, mintime, maxtime):
        '''Require time in keyword value to be in specified range.

        Specified keyword should have values that are absolute times.
        Apply filter to _mjd keyword because query fails for non-mjd keywords.
        Input times may be JD, MJD, astropy Time, datetime, or ISO 8601 string.
        Examples:
            filter_by_timerange('date_obs', '2022-04-02', '2022-04-03')
            filter_by_timerange('date_beg', '2022-04-02T11:00:00', 59671.5)
            filter_by_timerange('date_beg', '2022-04-02 11:00:00', 2459672)
        '''
        if keyword.lower().endswith('_mjd'):
            kw = keyword
        else:
            kw = keyword + '_mjd'
        try:
            minmjd = mjd_from_time(mintime)
            maxmjd = mjd_from_time(maxtime)
        except ValueError as e:
            raise e.with_traceback(e.__traceback__)
        newfilter = {
            'paramName': str(kw),
            'values': [{'min': minmjd, 'max': maxmjd}]}
        self.filters.append(newfilter)

    def set_output_columns(self, column_names):
        '''Set output columns as specified.

        column_names: str or list of str
            Comma-separated column names or list of column names
        '''
        self.columns = []
        self.append_output_columns(column_names)

    def set_output_columns_to_default(self):
        '''Set list of output columns to default value.'''
        inst_configs = {
            'Fgs': 'lamp',
            'GuideStar': 'gdstarid, gs_order',
            'Miri': 'filter, coronmsk, lamp',
            'Nircam': 'module, channel, pupil, filter, coronmsk', 
            'Niriss': 'pupil, filter, lamp', 
            'Nirspec': 'filter, grating, msastate, lamp', 
            }
        self.columns = []
        self.append_output_columns('date_beg, obs_id, category, targname')
        if self.collection != 'GuideStar':
            self.append_output_columns('template, expripar, numdthpt')
        self.append_output_columns('apername')
        try:
            self.append_output_columns(inst_configs[self.collection])
        except KeyError:
            raise ValueError(
                f"unknown collection: {self.collection}\n"
                f"known collections: {' '.join(inst_configs)}")
        self.append_output_columns('exp_type, detector, subarray')
        self.append_output_columns('readpatt, nints, ngroups, duration')
        # self.append_output_columns('productLevel, filename')

    def set_output_columns_to_all(self):
        '''Specify that all columns ('*') should be output.'''
        self.columns = '*'

    def prepend_required_output_columns(self):
        '''Prepend required output columns and remove duplicates.

        Require fileSetName and detector columns to construct the CAOM obsid.
        Use list(dict.fromkeys()) to remove duplicates, while preserving order.
        '''
        if self.columns == '*':
            return
        required_columns = ['filename']
        columns = list(dict.fromkeys(
            required_columns + self.columns))
        self.set_output_columns(columns)

    def append_output_columns(self, column_names):
        '''Append one or more output column names to current list.

        column_names: str or list of str
            Comma-separated column names or list of column names
        '''
        try:
            names = column_names.split(',')
        except AttributeError:
            names = column_names
        for name in names:
            stripped = name.strip()
            if stripped not in self.columns:
                self.columns.append(stripped)

    def remove_output_columns(self, column_names):
        '''Remove one or more output column names from current list.

        column_names: str or list of str
            Comma-separated column names or list of column names
        '''
        try:
            names = column_names.split(',')
        except AttributeError:
            names = column_names
        for name in names:
            stripped = name.strip()
            if stripped in self.columns:
                self.columns.remove(stripped)

    def execute_query(self, convert_dates=True):
        '''Execute query by calling MAST service with specified parameters.'''
        if not self.filters:
            raise ValueError('add search filter(s) before executing query')
        self.prepend_required_output_columns()
        params = {
            'columns': ','.join(self.columns),
            'filters': self.filters}
        self._params = params
        self.result = Mast.service_request(self.service, params)
        if convert_dates:
            self.convert_dates()

    def convert_dates(self):
        '''Convert table values containing /Date()/ to datetime objects.'''
        if self.result is None:
            raise RuntimeError('execute query before parsing dates')
        if len(self.result) == 0:
            return
        for colname in self.result.colnames:
            values = list(self.result[colname].data.data)
            newval = [
                datetime.utcfromtimestamp(int(v[6:19]) / 1000)
                if isinstance(v, str) and len(v) == 21 and
                    v[:6] == '/Date(' and v[-2:] == ')/' and
                    v[6:19].isdigit()
                else v for v in values]
            if all([isinstance(v, datetime) for v in newval]):
                self.result[colname] = newval
            elif any([n != v and not np_isnan(v) for n, v
                    in zip(newval, values)]):
                self.result[colname] = [v.isoformat()[:-3] for v in newval]

    def get_caom_obsid(self):
        '''Get CAOM obsid for archive files returned by JWST filtered query.

        References:
            https://mast.stsci.edu/api/v0/pyex.html#MastCaomFilteredPy
            https://mast.stsci.edu/api/v0/_services.html#MastCaomFiltered
            https://mast.stsci.edu/api/v0/_c_a_o_mfields.html
        '''
        if self.result is None:
            raise RuntimeError('execute query before getting CAOM obsid')
        dataset = ['_'.join(f.split('_')[:-1]) for f in self.result['filename']]
        self._dataset = sorted(list(set(dataset)))
        service = 'Mast.Caom.Filtered'
        filters = [
            {'paramName': 'obs_collection', 'values': ['JWST']},
            {'paramName': 'instrument_name', 'values': [f'{self.collection}']},
            {'paramName': 'obs_id', 'values': self.dataset}]
        params = {'columns': 'obsid', 'filters': filters}
        caom_filtered_result = Mast.service_request(service, params)
        self._caom_obsid = ','.join(caom_filtered_result['obsid'].data)

    def get_caom_product_list(self):
        self.get_caom_obsid()
        self.caom_product_list = CaomProductList(self.caom_obsid)

    def browse(self, unique=True):
        '''Show unique query results in a browser window.'''
        if self.result is None:
            raise RuntimeError('execute query before trying to show result')
        if unique:
            unique_result = table_unique(self.result, keys='filename')
            unique_result.show_in_browser(jsviewer=True)
        else:
            self.result.show_in_browser(jsviewer=True)
        if self.caom_product_list:
            self.caom_product_list.browse(unique=unique)

class CaomProductList:
    '''Get list of CAOM products for one or more CAOM product group IDs.

    Examples:
        CaomProductList('71738577')
        CaomProductList('71738577, 71738600')
        CaomProductList(['71738577'])
        CaomProductList(['71738577', '71738600'])
        CaomProductList(71738577)
        CaomProductList([71738577])
        CaomProductList([71738577, 71738600])

    References:
        https://mast.stsci.edu/api/v0/pyex.html#MastCaomProductsPy
        https://mast.stsci.edu/api/v0/_services.html#MastCaomProducts
        https://mast.stsci.edu/api/v0/_productsfields.html
    '''
    def __init__(self, caom_obsid):
        self._obsid = self.parse_caom_obsid(caom_obsid)
        self.product_list = self.get_product_list()

    @property
    def obsid(self):
        '''Return CAOM obsid list as a comma-separated string.'''
        return self._obsid

    def parse_caom_obsid(self, caom_obsid):
        '''Parse input specification of one or more CAOM obsid.

        caom_obsid: str or int or iterable yielding those types
            specification of one or more CAOM obsid
        '''
        try:
            obsid = caom_obsid.split(',')
        except AttributeError:
            obsid = caom_obsid
        try:
            return ','.join([str(int(obsid))])
        except TypeError:
            pass
        try:
            return ','.join([str(int(i)) for i in obsid])
        except (TypeError, ValueError):
            raise TypeError(
                f'CAOM obsid must evaluate to integer(s): {caom_obsid}')

    def get_product_list(self):
        '''Get list of CAOM products for specified CAOM obsid.'''
        service = 'Mast.Caom.Products'
        params = {'obsid': self.obsid}
        return Mast.service_request(service, params)

    def browse(self, unique=True):
        '''Show product list in a browser window.'''
        columns = 'obsID,productFilename,size,calib_level,' \
            'productSubGroupDescription,productType,dataproduct_type'
        product_list = self.product_list[columns.split(',')]
        if unique:
            product_list = table_unique(product_list, keys='productFilename')
        product_list.show_in_browser(jsviewer=True)

def mjd_from_time(time):
    '''Return modified Julian date equivalent to input time specification.

    time: JD/MJD float, astropy Time, tz aware/naive datetime, or ISO 8601 str
          Treat float-compatible input as MJD or JD, de[ending on value.
          Assume astropy Time is UTC.
          Treat timezone-naive datetime object as UTC.
          Treat ISO 8601 string without timezone specification as UTC.
    '''
    # Input is JD or MJD as numeric or str.
    try:
        jd_or_mjd = float(time)
        if jd_or_mjd > 2400000.5:
            mjd = jd_or_mjd - 2400000.5
        else:
            mjd = jd_or_mjd
        return mjd
    except (TypeError, ValueError):
        pass
    # Input is astropy Time object
    if isinstance(time, Time):
        return time.mjd
    # Input is python datetime object.
    if isinstance(time, datetime):
        dt = time
    else:
        try:
            dt = datetime.fromisoformat(time)
        except (TypeError, ValueError):
            raise ValueError(f'unable to parse time specification: {time}')
    # If timezone was not specified, assume UTC.
    naive = dt.tzinfo is None or dt.tzinfo.utcoffset(dt) is None
    if naive:
        return Time(dt.replace(tzinfo=timezone.utc)).mjd
    else:
        return Time(dt).mjd

def get_mast_api_token(mast_api_token=None, prompt=False):
    '''Get MAST API token. Precedence is argument, environment, file, prompt.'''
    if not mast_api_token:
        mast_api_token = getenv('MAST_API_TOKEN')
    if not mast_api_token:
        path = Path.home() / '.mast_api_token'
        try:
            with open(path, 'r') as fp:
                lines = fp.read().splitlines()
                if len(lines) == 1:
                    mast_api_token = lines[0]
                else:
                    print('Ignoring ~/.mast_api_token, expected one line')
        except FileNotFoundError:
            pass
    if not mast_api_token and prompt:
        mast_api_token = getpass('Enter MAST API token: ')
    try:
        assert(mast_api_token)
        assert(isinstance(mast_api_token, str))
        assert(len(mast_api_token) == 32)
        assert(mast_api_token.isalnum())
        return mast_api_token
    except AssertionError as e:
        raise ValueError(f"MAST API token is not a string " \
            f"with 32 alphanumeric characters: '{mast_api_token}'")
